sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/ObjectAnnotationVisitor.java

```diff
@@ -0,0 +1,311 @@
+                        var beanProperty = properties.get(i);
+                        ExpressionDef propertyValue;
+                        // get property value
+                        if ((!beanProperty.hasAnnotation(ToString.Exclude.class)) && beanProperty.getReadMethod().isPresent()) {
+                            propertyValue = parameterDef.get(0).asVariable().invoke(
+                                beanProperty.getReadMethod().get().getSimpleName(),
+                                TypeDef.of(beanProperty.getType()),
+                                List.of()
+                            );
+                        } else {
+                            continue;
+                        }
+
+                        exp = exp.invoke("append", variableDef.type(),
+                                ExpressionDef.constant(beanProperty.getName() + "="))
+                            .invoke("append", variableDef.type(),
+                                (TypeDef.of(beanProperty.getType()).isArray()) ?
+                                    ClassTypeDef.of(Arrays.class).invokeStatic("toString",
+                                        TypeDef.of(String.class), List.of(propertyValue)) : propertyValue)
+                            .invoke("append", variableDef.type(),
+                                ExpressionDef.constant((i == properties.size() - 1) ? "]" : ", "));
+                    }
+
+                    return exp.invoke("toString", TypeDef.of(String.class)).returning();
+                })
+            );
+        classDefBuilder.addMethod(method);
+    }
+
+    /*
+    Creates an equals method with signature:
+        public static boolean BeanNameObject.equals(BeanName object1, Object object2)
+     */
+    private static void createEqualsMethod(ClassDef.ClassDefBuilder classDefBuilder, String objectName, List<PropertyElement> properties) {
+        MethodDef method = MethodDef.builder("equals")
+            .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
+            .returns(boolean.class)
+            .addParameter("instance", ClassTypeDef.of(objectName))
+            .addParameter("o", TypeDef.of(Object.class))
+            .build((self, parameterDef) -> {
+                VariableDef instance = parameterDef.get(0).asVariable();
+                VariableDef o = parameterDef.get(1).asVariable();
+                return StatementDef.multi(
+                    instance.asCondition(" == ", o)
+                        .asConditionIf(ExpressionDef.trueValue().returning()),
+                    o.isNull().asCondition(" || ",
+                        instance.invoke("getClass", ClassTypeDef.of("Class"))
+                            .asCondition(" != ", o.invoke("getClass", ClassTypeDef.of("Class"))))
+                        .asConditionIf(ExpressionDef.falseValue().returning()),
+                    o.cast(ClassTypeDef.of(objectName)).newLocal("other", variableDef -> {
+                        ExpressionDef exp = null;
```

Comment:
Avoid variable definitions on the top

Reply:
But this increases readability. I thought that is also important?